source/db/dffml_source_db/source.py

```diff
@@ -0,0 +1,143 @@
+
+    def convert_to_record(self, result):
+        modified_record = {
+            "key": "",
+            "data": {"features": {}, "prediction": {}},
+        }
+        for key, value in result.items():
+            if key.startswith("feature_"):
+                modified_record["data"]["features"][
+                    key.replace("feature_", "")
+                ] = value
+            elif ("_value" in key) or ("_confidence" in key):
+                target = key.replace("_value", "").replace("_confidence", "")
+                modified_record["data"]["prediction"][target] = {
+                    "value": result[target + "_value"],
+                    "confidence": result[target + "_confidence"],
+                }
+            else:
+                modified_record[key] = value
+        return Record(modified_record["key"], data=modified_record["data"])
+
+    async def record(self, key: str):
+        record = Record(key)
+        async with self.parent.db_impl() as db_ctx:
+            try:
+                row = await db_ctx.lookup(
+                    self.parent.config.table_name,
+                    cols=None,  # None turns into *. We want all rows
+                    conditions=[[Condition("key", "=", key)]],
+                ).__anext__()
+            except StopAsyncIteration:
+                # This would happen if there is no matching row, so the async generator reached the end
+                return record
+
+        if row is not None:
+            features = {}
+            predictions = {}
+            for key, value in row.items():
+                if key.startswith("feature_"):
+                    features[key.replace("feature_", "")] = value
+                elif "_value" in key:
+                    target = key.replace("_value", "")
+                    predictions[target] = {
+                        "value": row[target + "_value"],
+                        "confidence": row[target + "_confidence"],
+                    }
+            record.merge(
+                Record(
+                    row["key"],
+                    data={"features": features, "prediction": predictions},
+                )
+            )
+        return record
+
+
+@entrypoint("dbsource")
+class DbSource(BaseSource):
+    CONFIG = DbSourceConfig
+    CONTEXT = DbSourceContext
+    db_impl: BaseDatabase
+
+    def __init__(self, cfg: Type[BaseConfig]) -> None:
+        super().__init__(cfg)
+
+    async def __aenter__(self) -> "DbSource":
+        # Instantiate a subclass of BaseDatabase using "db_implementation" from DbSourceConfig
+        self.db_impl = None
+        if self.config.db_implementation == "sqlite":
+            self.db_impl = SqliteDatabase(self.config.db_config)
+        # TODO: Add more implementations. Is there a better way to do this?
+        else:
+            raise RuntimeError(
+                f"No implementation named {self.config.db_implementation} found for DbSource abstraction"
+            )
```

Comment:
If we switch to the single `db: BaseDatabase` in the config I think we can remove all this and replace it with:

```python
    async def __aenter__(self) -> "DbSource":
        self.db_impl = await self.config.db.__aenter__()
```